jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/ThreadLimitHandler.java

```diff
@@ -339,45 +344,38 @@ protected void process(CompletableFuture<Closeable> futurePermit)
         @Override
         public void demand(Runnable onContent)
         {
-            Runnable permittedDemand = () ->
-            {
-                // TODO need to consider if we already have a permit!
-                CompletableFuture<Closeable> futurePermit = _remote.acquire();
+            if (_onContent != null)
+                throw new IllegalStateException();
+            _onContent = Objects.requireNonNull(onContent);
+            super.demand(this::onContent);
+        }
 
-                if (futurePermit.isDone())
-                {
-                    try
-                    {
-                        onContent.run();
-                    }
-                    finally
-                    {
-                        getAndClose(futurePermit);
-                    }
-                }
-                else
-                {
-                    futurePermit.thenAccept(c ->
-                    {
-                        try
-                        {
-                            onContent.run();
-                        }
-                        finally
-                        {
-                            IO.close(c);
-                        }
-                    });
-                }
-            };
+        private void onContent()
+        {
+            Permit permit = _remote.acquire();
+            if (permit.isAllocated())
+                onPermittedContent(permit);
+            else
+                permit.whenAllocated(this::onPermittedContent);
+        }
 
-            super.demand(permittedDemand);
+        private void onPermittedContent(Permit permit)
+        {
+            try
+            {
+                _onContent.run();
```

Comment:
Should not `_onContent` be nulled out here, so that future calls to `demand()` won't throw?
Are we missing a test case?

Reply:
Indeed. Bug and missing test.

Reply:
I added a test that showed the bug... plus another one.
ThreadLimitHandler looks to be very under tested (and always has been).

Perhaps it needs a separate PR?  I think it is still better (or at least broken differently) than HEAD.